### DISCUSSION TITLE
使用 pnpm + workspace 有什么有优点？

### DISCUSSION THREAD

**Discussion by zahngying** [cloudmoonocus/fe-monitoring-platform#2] [Q&A] [answered]
monorepo

**Comment by cloudmoonocus** [comment, marked as answer]
`pnpm` 是一个快速的、节省磁盘空间的包管理器，用于 JavaScript 和 Node.js 应用程序。而 `workspace` 是 `pnpm`（以及其他包管理器如 Yarn 和 npm）中的一个功能，它允许您在单个仓库中管理多个包（项目），这种方式通常被称为 monorepo（单一仓库）管理。

使用 `pnpm` 结合 `workspace` 有以下优点：

1. **节省磁盘空间**：`pnpm` 使用硬链接和符号链接的策略来管理 node_modules，这意味着如果多个项目使用相同版本的包，它们将共享相同的物理副本，而不是为每个项目复制一份。

2. **提高安装速度**：由于 `pnpm` 重用已下载的包，因此在多个项目中安装相同的依赖项时，可以显著减少安装时间。

3. **一致性和依赖性管理**：`workspace` 允许你在单个地方定义和协调多个项目的依赖关系，确保它们使用相同版本的共享依赖，这有助于保持项目间的一致性。

4. **集中化的工作流**：通过使用 `workspace`，你可以在单个仓库中执行构建、测试和发布等任务，这简化了管理多个相关项目的工作流程。

5. **交叉依赖**：在 monorepo 中，项目可能会相互依赖。`pnpm workspace` 允许本地交叉依赖关系的链接，这意味着你可以很容易地在本地开发时共享和测试代码，而无需发布到 npm。

6. **原子性操作**：在 monorepo 中执行操作（如安装、更新或删除依赖项）时，`pnpm` 会考虑到所有工作空间的包，从而确保一致性，并提供更原子化的操作。

7. **过滤器**：`pnpm workspace` 提供了过滤器机制，使你可以针对特定的子项目执行命令，而不是整个仓库，这在大型仓库中非常有用。

8. **可扩展性**：随着项目的增长，`pnpm workspace` 提供了良好的可扩展性，可以轻松地添加新的包或服务到你的 monorepo 中。

9. **更好的持续集成（CI）性能**：由于 `pnpm` 重用已下载的包，并且可以针对特定的更改执行命令，这可以帮助减少 CI 运行的时间和资源消耗。

10. **简化版本控制**：将所有相关项目保存在同一个版本控制系统下，可以简化团队的协作和代码审查流程。

`pnpm` 结合 `workspace` 提供了一种高效而强大的方式来管理和维护大型、复杂的 JavaScript 和 Node.js 项目，特别是对于采用 monorepo 方法的团队来说。
